tiledb/sm/query/readers/aggregators/min_max_aggregator.cc

```diff
@@ -0,0 +1,457 @@
+        throw MinMaxAggregatorStatusException(
+            "Fixed size min/max aggregates fixed buffer should be for one "
+            "element only.");
+      }
+    } else {
+      if (result_buffer.original_buffer_size_ !=
+          ComparatorAggregatorBase<T>::field_info_.cell_val_num_) {
+        throw MinMaxAggregatorStatusException(
+            "Fixed size min/max aggregates fixed buffer should be for one "
+            "element only.");
+      }
+    }
+  }
+
+  bool exists_validity = result_buffer.validity_vector_.buffer();
+  if (ComparatorAggregatorBase<T>::field_info_.is_nullable_) {
+    if (!exists_validity) {
+      throw MinMaxAggregatorStatusException(
+          "Min/max aggregates for nullable attributes must have a validity "
+          "buffer.");
+    }
+
+    if (*result_buffer.validity_vector_.buffer_size() != 1) {
+      throw MinMaxAggregatorStatusException(
+          "Min/max aggregates validity vector should be for one element only.");
+    }
+  } else {
+    if (exists_validity) {
+      throw MinMaxAggregatorStatusException(
+          "Min/max aggregates for non nullable attributes must not have a "
+          "validity buffer.");
+    }
+  }
+}
+
+template <typename T, typename Op>
+void ComparatorAggregator<T, Op>::aggregate_data(AggregateBuffer& input_data) {
+  optional<T> res{nullopt};
+
+  if (input_data.is_count_bitmap()) {
+    res = min_max<uint64_t>(input_data);
+  } else {
+    res = min_max<uint8_t>(input_data);
+  }
+
+  {
+    // This might be called on multiple threads, the final result stored in
+    // value_ should be computed in a thread safe manner.
+    std::unique_lock lock(value_mtx_);
+    if (res.has_value() &&
+        (ComparatorAggregatorBase<T>::value_ == std::nullopt ||
+         op_(res.value(), ComparatorAggregatorBase<T>::value_.value()))) {
+      ComparatorAggregatorBase<T>::value_ = res.value();
+    }
+
+    if (ComparatorAggregatorBase<T>::field_info_.is_nullable_ &&
+        res.has_value()) {
+      ComparatorAggregatorBase<T>::validity_value_ = 1;
+    }
+  }
+}
+
+template <typename T, typename Op>
+void ComparatorAggregator<T, Op>::copy_to_user_buffer(
+    std::string output_field_name,
+    std::unordered_map<std::string, QueryBuffer>& buffers) {
+  ComparatorAggregatorBase<T>::copy_to_user_buffer(output_field_name, buffers);
+}
+
+template <typename T, typename Op>
+template <typename BITMAP_T>
+optional<T> ComparatorAggregator<T, Op>::min_max(AggregateBuffer& input_data) {
+  optional<VALUE_T> value;
+  auto fixed_data = input_data.fixed_data_as<void*>();
+  auto var_data = ComparatorAggregatorBase<T>::field_info_.var_sized_ ?
+                      input_data.var_data() :
+                      nullptr;
+
+  // Run different loops for bitmap versus no bitmap and nullable versus non
+  // nullable. The bitmap tells us which cells was already filtered out by
+  // ranges or query conditions.
+  const uint64_t max_cell = input_data.includes_last_var_cell() ?
```

Comment:
Why do we need this `last_var_cell` check inside of the implementation? This seems to be too closely tied to the details of the storage format. Aggregate functions should able to run on a set of spans with no reference to details about the particular storage layout (such spans might implicitly be backed by tiles, but are not inherently).

Reply:
The tile offsets buffers doesn't include the last offset. This PR will not implement the change to add the extra offset at unfiltering... But once this change is implemented, this will go away.